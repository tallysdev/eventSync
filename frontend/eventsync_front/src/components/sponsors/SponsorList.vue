<template>
  <v-row justify="center">
    <v-col cols="12" lg="10">
      <h2 class="text-h4 text-start mb-6 font-weight-bold">Patrocinadores</h2>
      <template v-if="loading">
        <v-progress-circular indeterminate color="primary"></v-progress-circular>
      </template>
      <template v-else-if="sponsors.length === 0">
        <v-alert type="info" color="info" variant="tonal">
          <b>Nenhum patrocinador encontrado</b>
        </v-alert>
      </template>
      <v-row v-else>
        <v-col cols="12" md="6" v-for="sponsor in sponsors" :key="sponsor.id">
          <v-card>
            <v-card-title>{{ sponsor.name }}</v-card-title>
            <v-card-subtitle>{{ sponsor.email }}</v-card-subtitle>
            <v-card-text>{{ sponsor.description }}</v-card-text>
            <v-avatar size="100">
              <v-img :src="sponsor.logo"></v-img>
            </v-avatar>
          </v-card>
        </v-col>
      </v-row>
      <v-row justify="center" class="mt-4">
        <v-col cols="auto">
          <v-pagination
            v-model="currentPage"
            :length="totalPages"
            @input="fetchSponsors"
          ></v-pagination>
        </v-col>
      </v-row>
      <v-btn @click="openDialog" color="primary" class="mt-4">
        <b> Adicionar Patrocinador </b>
      </v-btn>
      <SponsorForm :dialog="dialog" @update:dialog="dialog = $event" @sponsor-added="onSponsorAdded" />
    </v-col>
    <v-snackbar v-model="snackbar" :timeout="4000" top :color="snackbarColor" p>
      {{ snackbarMessage }}
    </v-snackbar>
  </v-row>
</template>

<script setup lang="ts">
import { fetchSponsors } from '@/services/sponsorService'
import { ref, onMounted, watch } from 'vue'
import { type Sponsor } from '@/types/sponsor'
import SponsorForm from './SponsorForm.vue'

const sponsors = ref<Sponsor[]>([])
const currentPage = ref(1)
const totalPages = ref(1)
const itemsPerPage = 6
const dialog = ref(false)
const loading = ref(false)
const errorMessage = ref<string | null>(null)
const snackbar = ref(false)
const snackbarMessage = ref('')
const snackbarColor = ref('')

const openDialog = () => {
  dialog.value = true
}

const fetchSponsorsData = async () => {
  loading.value = true
  errorMessage.value = null

  const loadingTimeout = setTimeout(() => {
    loading.value = false
    errorMessage.value = 'O tempo de carregamento expirou. Tente novamente.'
  }, 10000) // 10 seconds timeout

  try {
    const response = await fetchSponsors(currentPage.value, itemsPerPage)
    sponsors.value = response.data.results
    totalPages.value = Math.ceil(response.data.count / itemsPerPage)
    clearTimeout(loadingTimeout)
  } catch (error) {
    console.error('Error fetching sponsors:', error)
  } finally {
    loading.value = false
  }
}

onMounted(() => {
  fetchSponsorsData()
})

watch(currentPage, () => {
  fetchSponsorsData()
})

const onSponsorAdded = (message: string) => {
  snackbarMessage.value = message
  snackbarColor.value = 'success'
  snackbar.value = true
  fetchSponsorsData()
}
</script>

<style scoped></style>
