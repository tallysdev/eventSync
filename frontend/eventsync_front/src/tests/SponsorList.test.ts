import { mount } from '@vue/test-utils'
import { describe, it, expect, vi, beforeEach } from 'vitest'
import SponsorList from '@/components/sponsors/SponsorList.vue'

// Mock the service
vi.mock('@/services/sponsorService', () => ({
  fetchSponsors: vi.fn()
}))

describe('SponsorList.vue', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('renders correctly', () => {
    const wrapper = mount(SponsorList)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('opens the dialog when the Add Sponsor button is clicked', async () => {
    // Arrange
    const wrapper = mount(SponsorList)

    // Act
    await wrapper.find('v-btn').trigger('click')

    // Assert
    const dialog = (wrapper.vm as any).dialog
    expect(dialog.value).toBe(true)
  })
})
