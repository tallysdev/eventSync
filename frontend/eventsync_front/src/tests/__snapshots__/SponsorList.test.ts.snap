// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SponsorList.vue > renders correctly 1`] = `
"<v-row justify="center">
  <v-col cols="12" lg="10">
    <h2 class="text-h4 text-start mb-6 font-weight-bold">Patrocinadores</h2>
    <v-alert type="info" color="info" variant="tonal"><b>Nenhum patrocinador encontrado</b></v-alert>
    <v-row justify="center" class="mt-4">
      <v-col cols="auto">
        <v-pagination modelvalue="1" length="1"></v-pagination>
      </v-col>
    </v-row>
    <v-btn color="primary" class="mt-4"><b> Adicionar Patrocinador </b></v-btn>
    <v-dialog model-value="false" max-width="600">
      <v-card>
        <v-card-title><span class="headline">Novo Patrocinador</span></v-card-title>
        <v-card-text>
          <v-form modelvalue="false" lazy-validation="">
            <v-text-field modelvalue="" label="Nome" rules="(value) => {
  if (!value) {
    return &quot;Campo obrigatório&quot;;
  }
  return true;
}" required=""></v-text-field>
            <v-text-field modelvalue="" label="Email" rules="(value) => {
  if (!value) {
    return &quot;Campo obrigatório&quot;;
  }
  return true;
},(value) => {
  if (!value) {
    return &quot;Email é necessário&quot;;
  }
  const emailRegex = /^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$/;
  if (!emailRegex.test(value)) {
    return &quot;Por favor, insira um email válido&quot;;
  }
  return true;
}" required=""></v-text-field>
            <v-text-field modelvalue="" label="Telefone" rules="(value) => {
  if (!value) {
    return &quot;Campo obrigatório&quot;;
  }
  return true;
}" required=""></v-text-field>
            <v-textarea modelvalue="" label="Descrição" rules="(value) => {
  if (!value) {
    return &quot;Campo obrigatório&quot;;
  }
  return true;
}" required=""></v-textarea>
            <v-file-input modelvalue="" label="Logo" rules="(value) => {
  if (!value) {
    return &quot;Campo obrigatório&quot;;
  }
  return true;
}" required=""></v-file-input>
          </v-form>
          <!--v-if-->
        </v-card-text>
        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn color="blue darken-1" text="">Cancelar</v-btn>
          <v-btn color="blue darken-1" text="" disabled="true">Salvar</v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </v-col>
  <v-snackbar modelvalue="false" timeout="4000" top="" color="" p=""></v-snackbar>
</v-row>"
`;
